download/buffer.go

```diff
@@ -0,0 +1,115 @@
+package download
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"time"
+
+	"github.com/dustin/go-humanize"
+	"github.com/spf13/viper"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/replicate/pget/optname"
+)
+
+var (
+	fileSize int64
+)
+
+func getRemoteFileSize(url string) (string, int64, error) {
+	// TODO: this needs a retry
+	resp, err := http.DefaultClient.Head(url)
+	if err != nil {
+		return "", int64(-1), err
+	}
+	defer resp.Body.Close()
+	trueUrl := resp.Request.URL.String()
+	if trueUrl != url {
+		fmt.Printf("Redirected to %s\n", trueUrl)
+	}
+
+	fSize := resp.ContentLength
+	if fSize <= 0 {
+		return "", int64(-1), fmt.Errorf("unable to determine file size")
+	}
+	fileSize = fSize
+	return trueUrl, fileSize, nil
+}
+
+func FileToBuffer(url string) (*bytes.Buffer, int64, error) {
+	maxConcurrency := viper.GetInt(optname.Concurrency)
+
+	trueURL, fileSize, err := getRemoteFileSize(url)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	chunkSize := viper.GetInt64(optname.MinimumChunkSize)
+	// not more than one connection per min chunk size
+	concurrency := int(math.Ceil(float64(fileSize) / float64(chunkSize)))
+
+	if concurrency > maxConcurrency {
+		concurrency = maxConcurrency
+		chunkSize = fileSize / int64(concurrency)
+	}
+	if viper.GetBool(optname.Verbose) {
+		fmt.Printf("Downloading %s bytes with %d connections (chunk-size = %s)\n", humanize.Bytes(uint64(fileSize)), concurrency, humanize.Bytes(uint64(chunkSize)))
+	}
+
+	errGroup, ctx := errgroup.WithContext(context.Background())
+
+	data := make([]byte, fileSize)
```

Comment:
we'll need to put a lot of thought into this line

Reply:
Part of my plan is to evolve so we have a few new ways to download and we can keep the default as is. This buffer creation bit is just the current way we do things. I def. want something better than "ALLOCATE ALL THE RAM" :)